src/gui/qgsmapcanvas.cpp

```diff
@@ -202,6 +202,7 @@ QgsMapCanvas::QgsMapCanvas( QWidget * parent, const char *name )
   mCurrentLayer = NULL;
   mMapOverview = NULL;
   mMapTool = NULL;
+  mOldCursor = cursor();
```

Comment:
I guess there is a conceptual problem about this:

> Due to the fact that maptools don’t restore cursor, the previous cursor shape is kept in mOldCursor.

I don't see a reason why maptools should not restore the cursor (like everything else in the application does)
Qt has a cursor stack:

 * you want to change the cursor: push a change
 * you want to restore the curstor: pop the change

But since only the maptool knows if (and how many times) it pushed something onto the cursor stack only the maptool itself can pop that again.